### PR TITLE
High-level API to build dictionaries

### PR DESCRIPTION
**Pull request by moreal**
It's related with #6 but this includes only *proposal 1*.
it's to prevent too bigger pull-request size.

### CODE REVIEW

File: Bencodex.Tests/Types/DictionaryTest.cs [hunk 1]
```diff
@@ -74,5 +75,73 @@ public void Equality()
...
         }
+
+        [Fact]
+        public void SetItem()
+        {
+            var dictionary = Dictionary.Empty
+                .SetItem("name", "Alex")
+                .SetItem("age", 1337)
+                .SetItem("address", new byte[] { 1, 2, 3, 4 })
+                .SetItem("happiness", true);
```
Comment (dahlia) [inline Bencodex.Tests/Types/DictionaryTest.cs]: It would be good if it can take `IEnumerable<IValue>` as well.

File: Bencodex/Types/Dictionary.cs [hunk 2]
```diff
@@ -11,8 +11,8 @@ namespace Bencodex.Types
 {
     public struct Dictionary :
         IValue,
-        IEquatable<IImmutableDictionary<IKey, IValue>>,
-        IImmutableDictionary<IKey, IValue>
```
Comment (dahlia) [inline Bencodex/Types/Dictionary.cs]: Could you explain why did you make it to no more implement `IImmutableDictionary<IKey, IValue>`?  Making all `IValue` types immutable was an intention.

Comment (moreal) [reply Bencodex/Types/Dictionary.cs]: It was because return type of `Dictionary.SetItem(IKey, IValue)`can't be replaced to `Dictionary` from `IImmutableDictionary<IKey, IValue>`. Still, `Dictionary`'s Immutability is still being maintained.

Comment (dahlia) [reply Bencodex/Types/Dictionary.cs]: Doesn't C# allow to overload methods returning different types if they have different types of parameters?  See also: <https://dotnetfiddle.net/vXejQh>.

Comment (moreal) [reply Bencodex/Types/Dictionary.cs]: No, it's allowed. I just thought about it, but I cancelled it because I was not sure that it's right way. Anyway now, it seems better to overload methods returning different types, so I will apply it.

File: Bencodex.Tests/Types/DictionaryTest.cs [hunk 3]
```diff
@@ -2,6 +2,7 @@
 using System.Collections.Generic;
 using Bencodex.Types;
 using Xunit;
+using Boolean = Bencodex.Types.Boolean;
```
Comment (dahlia) [inline Bencodex.Tests/Types/DictionaryTest.cs]: Was this intended?  I guess you were intended to use qualified names instead, as you already did it in the below code!

File: Bencodex.Tests/Types/DictionaryTest.cs [hunk 4]
```diff
@@ -74,5 +75,89 @@ public void Equality()
...
+            Assert.Throws<InvalidCastException>(() => dictionary.GetValue<Boolean>("binary"));
+
+            Assert.Equal((Bencodex.Types.Boolean) true, (Bencodex.Types.Boolean) dictionary["boolean"]);
+            Assert.Equal((Bencodex.Types.Boolean) true, dictionary.GetValue<Boolean>("boolean"));
+            Assert.Throws<InvalidCastException>(() => dictionary.GetValue<Integer>("boolean"));
+            Assert.Throws<InvalidCastException>(() => dictionary.GetValue<Binary>("boolean"));
+            Assert.Throws<InvalidCastException>(() => dictionary.GetValue<Text>("boolean"));
+
+            Assert.Equal(new IValue[] { (Text)"bar", (Integer)1337 }, (Bencodex.Types.List) dictionary["list"]);
+            Assert.Equal(new IValue[] { (Text)"bar", (Integer)1337 }, dictionary.GetValue<Bencodex.Types.List>("list"));
```
Comment (dahlia) [inline Bencodex.Tests/Types/DictionaryTest.cs]: Some lines like this seem too long.  FYI at the time I started this project, I maintained every line is not over 80 characters.  I agree that it is not a good standard for C# projects, but anyway could you make the code style consistent to what it had been?

File: CHANGES.md [hunk 5]
```diff
@@ -6,6 +6,47 @@ Version 0.3.0
 
 To be released.
 
+ -  `Dictionary.SetItem` became to have more overloads.  [#7]
```
Comment (dahlia) [inline CHANGES.md]:
```suggestion
 -  `Bencodex.Types.Dictionary.SetItem()` became to have more overloads.  [[#7]]
```

File: CHANGES.md [hunk 6]
```diff
@@ -6,6 +6,47 @@ Version 0.3.0
...
+     -  (string, bool)
+     -  (string, IEnumerable<IValue>)
+     -  (byte[], IValue)
+     -  (byte[], string)
+     -  (byte[], byte[])
+     -  (byte[], long)
+     -  (byte[], ulong)
+     -  (byte[], bool)
+     -  (byte[], IEnumerable<IValue>)
+ -  `Dictionary.Add` became to have more overloads.  [#7]
```
Comment (dahlia) [inline CHANGES.md]:
```suggestion
 -  `Bencodex.Types.Dictionary.Add()` became to have more overloads.  [[#7]]
```

File: CHANGES.md [hunk 7]
```diff
@@ -6,6 +6,47 @@ Version 0.3.0
...
+      -  (byte[], byte[])
+      -  (byte[], long)
+      -  (byte[], ulong)
+      -  (byte[], bool)
+      -  (byte[], IEnumerable<IValue>)
+ -  Added `Dictionary[string]` indexer. [#7]
+ -  Added `Dictionary[byte[]]` indexer. [#7]
+
+[#7]: https://github.com/planetarium/bencodex.net/pull/7
+
```
Comment (dahlia) [inline CHANGES.md]:
```suggestion


```

File: CHANGES.md [hunk 8]
```diff
@@ -6,6 +6,47 @@ Version 0.3.0
...
+      -  (string, bool)
+      -  (string, IEnumerable<IValue>)
+      -  (byte[], IValue)
+      -  (byte[], string)
+      -  (byte[], byte[])
+      -  (byte[], long)
+      -  (byte[], ulong)
+      -  (byte[], bool)
+      -  (byte[], IEnumerable<IValue>)
+ -  Added `Dictionary[string]` indexer. [#7]
```
Comment (dahlia) [inline CHANGES.md]:
```suggestion
 -  Added `Bencodex.Types.Dictionary[string]` indexer. [#7]
```

File: CHANGES.md [hunk 9]
```diff
@@ -6,6 +6,47 @@ Version 0.3.0
...
+ -  Added `Dictionary[byte[]]` indexer. [#7]
```
Comment (dahlia) [inline CHANGES.md]:
```suggestion
 -  Added `Bencodex.Types.Dictionary[byte[]]` indexer. [#7]
```

File: CHANGES.md [hunk 10]
```diff
@@ -6,6 +6,47 @@ Version 0.3.0
...
+     -  (string, ulong)
+     -  (string, bool)
+     -  (string, IEnumerable<IValue>)
+     -  (byte[], IValue)
+     -  (byte[], string)
+     -  (byte[], byte[])
+     -  (byte[], long)
+     -  (byte[], ulong)
+     -  (byte[], bool)
+     -  (byte[], IEnumerable<IValue>)
```
Comment (dahlia) [inline CHANGES.md]: Could you format these items as monospace?  Plus, it would be better to mention that these overloads return `Bencodex.Types.Dictionary` instead of `IImmutableDictionary<IKey, IValue>`.

File: Bencodex/Types/Dictionary.cs [hunk 11]
```diff
@@ -152,12 +394,8 @@ public IEnumerable<byte[]> EncodeIntoChunks()
...
             yield return new byte[1] { 0x64 }; // 'd'
             IEnumerable<ValueTuple<byte?, byte[], IValue>> rawPairs =
-                from pair in this
-                select (
-                    pair.Key.KeyPrefix,
-                    pair.Key.EncodeAsByteArray(),
-                    pair.Value
-                );
+                Value.Select(pair => (pair.Key.KeyPrefix,
+                    pair.Key.EncodeAsByteArray(), pair.Value));
```
Comment (dahlia) [inline Bencodex/Types/Dictionary.cs]: Is this change necessary?

File: Bencodex/Types/Dictionary.cs [hunk 12]
```diff
@@ -11,7 +11,7 @@ namespace Bencodex.Types
 {
     public struct Dictionary :
         IValue,
-        IEquatable<IImmutableDictionary<IKey, IValue>>,
+        IEquatable<Dictionary>,
```
Comment (dahlia) [inline Bencodex/Types/Dictionary.cs]: Could you explain this change?

File: Bencodex/Types/Dictionary.cs [hunk 13]
```diff
@@ -185,7 +423,7 @@ public IEnumerable<byte[]> EncodeIntoChunks()
         [Pure]
         public override string ToString()
         {
-            IEnumerable<string> pairs = this.Select(
+            IEnumerable<string> pairs = Value.Select(
```
Comment (dahlia) [inline Bencodex/Types/Dictionary.cs]: Is this change necessary?

File: Bencodex/Types/Dictionary.cs [hunk 14]
```diff
@@ -108,26 +342,34 @@ public bool TryGetKey(IKey equalKey, out IKey actualKey)
...
         }
 
-        bool IEquatable<IImmutableDictionary<IKey, IValue>>.Equals(
-            IImmutableDictionary<IKey, IValue> other
+        bool IEquatable<Dictionary>.Equals(
+            Dictionary other
         )
         {
-            if (this.LongCount() != other.LongCount()) return false;
+            if (Value.LongCount() != other.Value.LongCount()) return false;
```
Comment (dahlia) [inline Bencodex/Types/Dictionary.cs]: Is this change necessary?

Comment (moreal) [reply Bencodex/Types/Dictionary.cs]: It seems there are lines remained although `Dictionary` became to implement `IImmutableDictionary<IKey, IValue>` again. I will see the lines again and apply your reviews.

File: Bencodex.Tests/Types/DictionaryTest.cs (hunk 4 above)
Comment (moreal) [reply Bencodex.Tests/Types/DictionaryTest.cs]: I agree about your opinion. Also it seems to need to set up editorcofnig.

File: Bencodex.Tests/Types/DictionaryTest.cs [hunk 15]
```diff
@@ -74,5 +74,144 @@ public void Equality()
...
+                dictionary.GetValue<Bencodex.Types.Boolean>("boolean"));
+            Assert.Throws<InvalidCastException>(
+                () => dictionary.GetValue<Integer>("boolean"));
+            Assert.Throws<InvalidCastException>(
+                () => dictionary.GetValue<Binary>("boolean"));
+            Assert.Throws<InvalidCastException>(
+                () => dictionary.GetValue<Text>("boolean"));
+
+            Assert.Equal(
+                new IValue[] { (Text)"bar", (Integer)1337 },
```
Comment (dahlia) [inline Bencodex.Tests/Types/DictionaryTest.cs]: Code style seems inconsistent: `(T) v` vs `(T)v`, but it's okay, we could introduce lints & style checkers in the separate pull request.

File: CHANGES.md [hunk 16]
```diff
@@ -6,6 +6,55 @@ Version 0.3.0
 
 To be released.
 
+ -  `Bencodex.Types.Dictionary.SetItem()` became
+    to have more overloads.  [[#7]]
+     -  The type `Bencodex.Types.Dictionary.SetItem()` returns became
+        `Bencodex.Types.Dictionary`, not `IImmutableDictionary<IKey, IValue>`,
+        except for `Bencodex.Types.Dictionary.SetItem(IKey, IValue)`.
```
Comment (dahlia) [inline CHANGES.md]: How about this?

> Added overloads, which is listed below, return `Bencodex.Types.Dictionary` instead of `IImmutableDictionary<IKey, IValue>`.  Note that existing `SetItem(IKey, IValue)` method which implements `IImmutableDictionary<IKey, IValue>` is still remained as it had been.

File: Bencodex.Tests/Types/DictionaryTest.cs (hunk 15 above)
Comment (moreal) [reply Bencodex.Tests/Types/DictionaryTest.cs]: I replaced them like `(T) v`, to maintain consistent already exists in this project.
After this pull-request is merged, I will open the pull-request you mentioned.
